File: lib/talker.py
from .asset.directory import Directory


class Talker:
    """Talker object prints colorful messages to the console"""

    @staticmethod
    def arrow_header_info(title, info, new_line=False) -> None:
        Talker._print(f'{title}: \n'
                      f'--->| {info} |<---', new_line)

    @staticmethod
    def arrow_info(title, info, new_line=False) -> None:
        message = f'{title} --> {info}'
        Talker._print(message, new_line)

    @staticmethod
    def list_info(items, title, new_line=False) -> None:
        for x in items:
            Talker.info(title, x)
        if new_line:
            print()

    @staticmethod
    def list_dir_info(dirs, title, new_line=False) -> None:
        links = [directory.url for directory in dirs if isinstance(directory, Directory)]
        Talker.list_info(links, title, new_line)

    @staticmethod
    def info(title, info, new_line=False) -> None:
        message = f'{title}: {info}'
        Talker._print(message, new_line)

    @staticmethod
    def header(title, new_line=False) -> None:
        message = f'------{title}------'
        Talker._print(message, new_line)

    @staticmethod
    def loading(title, dots=5, new_line=False) -> None:
        dots = '.' * dots
        message = f'{title}{dots}'
        Talker._print(message, new_line)

    @staticmethod
    def complete(title, new_line=False) -> None:
        message = f'*** {title} ***'
        Talker._print(message, new_line)

    @staticmethod
    def warning(title, new_line=False) -> None:
        message = f'!- {title} -!'
        Talker._print(message, new_line)

    @staticmethod
    def divider() -> None:
        message = f"================================================="
        Talker._print(message, True)

    @staticmethod
    def _print(message, new_line) -> None:
        if new_line:
            message += '\n'
        print(message)

    @staticmethod
    def current_directory(verbose, url, depth_level) -> None:
        Talker.header("Current Directory")
        if verbose:
            Talker.arrow_info("level", depth_level)
        Talker.arrow_info("url", url, True)

    @staticmethod
    def file_stats(dirs, files) -> None:
        Talker.header("Directories")
        Talker.list_dir_info(dirs, 'Dir')
        Talker.arrow_info("Total", len(dirs), True)
        Talker.header("Files")
        Talker.list_info(files, 'File')
        Talker.arrow_info("Total", len(files))
        Talker.divider()

File: app/src/main/java/xyz/zimtools/zyod/Talker.java
package xyz.zimtools.zyod;

import xyz.zimtools.zyod.assets.Directory;
import xyz.zimtools.zyod.assets.ODUrl;

import java.util.List;

/**
 * Prints messages to the console
 */
public final class Talker {
    public static void fileStats(List<Directory> dirs, List<ODUrl> files) {
        Talker.header("Directories", false);
        Talker.dirListInfo(dirs, "Dir", false);
        Talker.arrowInfo("Total", String.valueOf(dirs.size()), true);
        Talker.header("Files", false);
        Talker.fileListInfo(files, "File", false);
        Talker.arrowInfo("Total", String.valueOf(files.size()), false);
        Talker.divider();
    }

    public static void currentDirectory(Directory directory, boolean isVerbose) {
        Talker.header("Current Directory", false);
        if (isVerbose) {
            Talker.arrowInfo("level", String.valueOf(directory.getDepthLevel()), false);
        }
        Talker.arrowInfo("url", directory.toString(), true);
    }

    public static void dirListInfo(List<Directory> dirs, String title, boolean newLine) {
        for (Directory l : dirs) {
            Talker.info(title, l.toString(), newLine);
        }
    }

    public static void fileListInfo(List<ODUrl> files, String title, boolean newLine) {
        for (ODUrl f : files) {
            Talker.info(title, f.toString(), newLine);
        }
    }

    public static void info(String title, String info, boolean newLine) {
        Talker.println(String.format("%s: %s", title, info), newLine);
    }

    public static void arrowHeaderInfo(String title, String info, boolean newLine) {
        Talker.println(String.format("%s: %n" +
                "--->| %s |<---", title, info), newLine);
    }

    public static void arrowInfo(String title, String info, boolean newLine) {
        Talker.println(String.format("%s --> %s", title, info), newLine);
    }

    private static void println(String message, boolean newLine) {
        if (newLine) {
            message += "\n";
        }
        System.out.println(message);
    }

    public static void loading(String title, boolean newLine) {
        String dots = ".....";
        String message = title + dots;
        println(message, newLine);
    }

    public static void complete(String title, boolean newLine) {
        String stars = "***";
        String message = String.format("%1s %2s %1s", stars, title, stars);
        println(message, newLine);
    }

    public static void warning(String title, boolean newLine) {
        String message = String.format("!- %s -!", title);
        println(message, newLine);
    }

    public static void divider() {
        String message = "=================================================";
        println(message, true);
    }

    public static void header(String title, boolean newLine) {
        String message = String.format("------%s------", title);
        Talker.println(message, newLine);
    }


}